nextcloud26: init at 26.0.0beta1

urandom2 opened the pull request:
###### Description of changes

Because nextcloud ships their prerelease versions on a different url, we are not parsing the version string to detect which path to use. We also enabled and validated this change via nixos module testing.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

urandom2 committed:
nextcloud26: init at 26.0.0beta1

Because nextcloud ships their prerelease versions on a different url, we
are not parsing the version string to detect which path to use. We also
enabled and validated this change via nixos module testing.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1760

onny commented:
Is it possible to already generate and include nextcloud26Packages?

urandom2 committed:
nextcloud26Packages: init at various

While we had to lie to generate.sh in order to get nc26 into the
running, and there are a lot of apps that have not released a compatible
version yet, for the ones that have, we are adding them alongside
nextcloud 26.0.0beta1.

urandom2 commented:
yep, added a new commit for that; it is very much not all there yet, but most of the core stuff is, and I had to bodge `generate.sh` with a `26.0.0` version manually, since it appears to be using the wrong nixpkgs, and did not like `26.0.0beta1`

onny commented:
Result of `nixpkgs-review pr 212912` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nextcloud26</li>
  </ul>
</details>

onny commented:
Tests fail to run for me :(
```
~/p/nixpkgs (master) [1]> nixpkgs-review pr 212912 -p nixosTests.nextcloud
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/212912/head:refs/nixpkgs-review/1
$ git worktree add /home/onny/.cache/nixpkgs-review/pr-212912-4/nixpkgs 6173948c03942a3eaff2e18c7375f4f79c3b1739
Preparing worktree (detached HEAD 6173948c039)
Updating files: 100% (33420/33420), done.
HEAD is now at 6173948c039 Merge pull request #212969 from r-ryantm/auto-update/python310Packages.google-cloud-language
$ git merge --no-commit --no-ff c71e752e7baaad90a1447426741a113b27a2219d
Auto-merging pkgs/top-level/all-packages.nix
Automatic merge went well; stopped before committing as requested
error: cannot coerce a set to a string

       at /nix/store/0zipgmy333i0x71xgriaa5ajz11waf1s-nixpkgs-review-2.7.0/lib/python3.10/site-packages/nixpkgs_review/nix/evalAttrs.nix:16:27:

           15|       else
           16|         builtins.tryEval "${pkg}";
             |                           ^
           17|   in rec {
nix --experimental-features nix-command --system x86_64-linux eval --json --impure --expr (import /nix/store/0zipgmy333i0x71xgriaa5ajz11waf1s-nixpkgs-review-2.7.0/lib/python3.10/site-packages/nixpkgs_review/nix/evalAttrs.nix { allowAliases = false; attr-json = /tmp/tmpwrgmy9qp; }) failed to run, /tmp/tmpwrgmy9qp was stored inspection
https://github.com/NixOS/nixpkgs/pull/212912 failed to build
```

urandom2 commented:
I could be wrong, but this looks like it could be an issue with `nixpkgs-review`, rebasing off the tip of master (what I just pushed), works fine and does not reproduce the failure. also `nix-build -A nixosTests.nextcloud` completes successfully.

Ma27 reviewed (commented):
What's the motivation to ship pre-releases? I'm somewhat hesitant to do that.

urandom2 commented:
I am interested in running this on my staging infrastructure prior to the full launch, to validate the new release. we have plenty of other prerelease packages with `unstable-2023-01-28` version strings, and you have to manually opt in, what is the harm?

onny commented:
> I could be wrong, but this looks like it could be an issue with `nixpkgs-review`, rebasing off the tip of `master` (what I just pushed), works fine and does not reproduce the failure. also `nix-build -A nixosTests.nextcloud` completes successfully.

Okay manually building the tests work fine, thank you :)

I'm interested to package Nextcloud 26 beta in order to better test and integrate upcoming versions for NixOS (I'm also doing some development for Nextcloud). Thank you for this work!

SuperSandro2000 commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -51,14 +52,19 @@ in {
+    version = "26.0.0beta1";
+    hash = "sha256-EfSfn0KjQzciHa3VcrDhGC/aZUw/KDjihXs+qVIcYX0=";
+  };
+
+  # tip: get hash with:
+  # nix hash to-sri --type sha256 $(curl https://download.nextcloud.com/server/releases/nextcloud-${version}.tar.bz2.sha256 | cut -d' ' -f1)
```
Or just use TOFU...

urandom2 replied:
that is just what I do, but I did not feel comfortable simply deleting the comment; happy to if you would rather.

SuperSandro2000 commented:
> I am interested in running this on my staging infrastructure prior to the full launch, to validate the new release.

You can also do that with overlays very easily. If not we should adjust the packaging of nextcloud.

> we have plenty of other prerelease packages with `unstable-2023-01-28` version strings, and you have to manually opt in, what is the harm?

That statement is wrong. ``unstable-XXXX-XX-XX`` is used if there are no release tags or the latest release is years in the past and we would need to cherry-pick 10s of patches to make it work.

Normally we don't really want to ship beta versions of softwares. They do not work great with our stable release workflow and require more maintenance and work. They are the exceptions and we should only do it if there is a strong reason for it.

I don't see one here, so I am fully with @Ma27 .

Also backporting this to 22.11 is wrong. I am going to revert that in #213226

SuperSandro2000 commented:
Also I am inclined to revert the addition of nextcloud26 beta, if @Ma27 agrees with that. Nextcloud does not allow to easily downgrade your database in case there are problems and it is not worth the extra effort.

urandom2 commented:
> You can also do that with overlays very easily. If not we should adjust the packaging of nextcloud.

yeah, doing this via an overlay is near impossible via the `let` usage of the `generic` function. are you recommending passing the current `generic` function through some symbol like `pkgs.nextcloud.mkDerivation`?

> That statement is wrong.` unstable-XXXX-XX-XX` is used if there are no release tags or the latest release is years in the past and we would need to cherry-pick 10s of patches to make it work.

considering how we package nextcloud, that is how I read the `nextcloud26` symbol, it is a derivation with "no release tags", and since we had others in nixpkgs, it seems fine; then once they cut the first tag, we can move over.

> Also backporting this to 22.11 is wrong. I am going to revert that in https://github.com/NixOS/nixpkgs/pull/213226

as such, I am mixed on this being "wrong", but I can see that there is not much benefit to the backport, and my actions there reflected my own interest and the perceived eagerness of the community via the actions of a core maintainer.

---

> Nextcloud does not allow to easily downgrade your database in case there are problems and it is not worth the extra effort.

that said, I would ask how much we concerned with people accidentally upgrading? I get that downgrades are impossible, but that this is a problem with loads more software than nextcloud. e.g. signal. also, there are a lot of things that even jumping between stable releases can break, and as you call out it is impossible to go back. as such, my upgrade process usually involves a pile of backups, and a well laid plan; I get that everybody may not, but there is only so much we can do?

if this change would serve to make nixos and nixpkgs worse by breaking lots of installations, could merge the package with a broken or unsupported platform flag? this is kind of a misuse of those flags, so maybe the right move is to advocate for the idea of a "beta/unstable" flag in derivation metadata?

urandom2 commented:
> Normally we don't really want to ship beta versions of softwares.

see `go_1_20`, which tracked `rc{1,2,3}`

SuperSandro2000 commented:
> yeah, doing this via an overlay is near impossible via the `let` usage of the `generic` function. are you recommending passing the current `generic` function through some symbol like `pkgs.nextcloud.mkDerivation`?

Yes, for example.



> and my actions there reflected my own interest

For that I just use my fork. I don't think we should backport most commits I cherry-picked in my branch especially the breaking ones.

https://github.com/NixOS/nixpkgs/compare/nixos-22.11...SuperSandro2000:nixpkgs:nixos-22.11
